test/deployer/s3_pusher_test.rb

```diff
@@ -0,0 +1,59 @@
+require "minitest/autorun"
+require "mocha/minitest"
+require "aws-sdk-s3"
+
+require "serverless-tools/deployer/s3_pusher"
+require "serverless-tools/deployer/function_config"
+
+module ServerlessTools::Deployer
+  describe "S3Pusher" do
+    let(:s3) { Aws::S3::Client.new(stub_responses: true) }
+    let(:config) do
+      FunctionConfig.new(name: "filename", bucket: "test", s3_archive_name: "function.zip")
+    end
+    let(:git) { mock }
+    let(:builder) { mock }
+
+    before do
+      git.stubs(:sha).returns("1234567890")
+      builder.stubs(:local_filename).returns("filename.zip")
+    end
+
+    let(:object) { mock }
+    let(:subject) { S3Pusher.new(client: s3, git: git, builder: builder) }
+    let(:expected) do
+      {
+        s3_bucket: "test",
+        s3_key: "/deployments/1234567890/filename/function.zip"
+      }
+    end
+
+    describe "#push" do
+      describe "when an object doesn't exist" do
+        before do
+          Aws::S3::Object.any_instance.stubs(:exists?).returns(false)
+          Aws::S3::Object.any_instance.expects(:upload_file).with("filename.zip")
+        end
+
+        it "uploads the file and returns the uploaded configuration" do
+          result = subject.push(config: config)
+          expected = {
+            s3_bucket: "test",
+            s3_key: "/deployments/1234567890/filename/function.zip"
+          }
```

Comment:
I think it would also be nice to make tests from both the `describe` blocks (`"when an object doesn't exist"` and `"when an object does exist"`) follow the same structure - at the moment one uses a `before` blocks to wrap stubs on `Aws::S3::Object` whether the other one doesn't.